### PR TITLE
CB-5868 No suitable subnet found as there were neither private nor an…

### PR DESCRIPTION
**Pull request by gergopapi2**

### CODE REVIEW

File: cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/service/subnetselector/SubnetSelectorStrategy.java [hunk 1]
```diff
@@ -1,55 +1,61 @@
...
-        }
-        if (subnetMetas.size() < getMinimumNumberOfSubnets()) {
-            error(String.format("There are not enough subnets in this network, found: %d, expected: %d.", subnetMetas.size(), getMinimumNumberOfSubnets()));
+        Optional<String> errorMessage = quickValidate(subnetMetas);
+        if (errorMessage.isPresent()) {
+            return new SubnetSelectionResult(errorMessage.get());
         }
+        SubnetSelectionResult selectionResult = selectInternal(subnetMetas);
+        logResult(selectionResult);
+        return selectionResult;
```
Comment (lacikaaa) [inline cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/service/subnetselector/SubnetSelectorStrategy.java]: `else` is missing

Comment (gergopapi2) [reply cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/service/subnetselector/SubnetSelectorStrategy.java]: What do you mean by else? You mean, fallback strategy? That you will find at SubnetIdProvider or at SubnetChooser, as they do different things on fallback (does this answer your question?)

Comment (lacikaaa) [reply cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/service/subnetselector/SubnetSelectorStrategy.java]: you have an if, and this is the else part. makes the code easier to read if you don't have hidden returns

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/subnet/selector/AzureSubnetSelectorService.java [hunk 2]
```diff
@@ -3,34 +3,38 @@
...
+        List<CloudSubnet> selectedSubnets = subnetSelectionParameters.isForDatabase()
                 ? new ArrayList<>(subnetMetas)
                 : List.of(subnetMetas.get(0));
+        return logResult(new SubnetSelectionResult(selectedSubnets));
     }
 
-    private List<CloudSubnet> logAndThrow(String message) {
-        LOGGER.debug(message);
-        throw new BadRequestException(message);
+    private SubnetSelectionResult logResult(SubnetSelectionResult subnetSelectionResult) {
```
Comment (lacikaaa) [inline cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/subnet/selector/AzureSubnetSelectorService.java]: a bit strange that logresult returns

File: environment/src/main/java/com/sequenceiq/environment/network/service/SubnetIdProvider.java [hunk 3]
```diff
@@ -30,9 +30,16 @@ public String provide(NetworkDto network, Tunnel tunnel, CloudPlatform cloudPlat
...
                 .networkConnector()
                 .selectSubnets(new ArrayList<>(network.getSubnetMetas().values()), SubnetSelectionParameters.builder().withTunnel(tunnel).build());
-        return selectedSubnets.get(0).getId();
+        CloudSubnet selectedSubnet = subnetSelectionResult.hasResult()
+                ? subnetSelectionResult.getResult().get(0)
+                : fallback(network);
+        return selectedSubnet.getId();
+    }
+
+    private CloudSubnet fallback(NetworkDto network) {
```
Comment (lacikaaa) [inline environment/src/main/java/com/sequenceiq/environment/network/service/SubnetIdProvider.java]: add a log here please

File: cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/service/subnetselector/SubnetSelectorStrategy.java (hunk 1 above)
Comment (gergopapi2) [reply cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/service/subnetselector/SubnetSelectorStrategy.java]: I am not sure I agree. I think this is opinion based, as some like single returns, whereas others prefer early returns. I personally think there is value in early returns as it avoids nesting and gives a clear flow to the code.

File: cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/subnet/selector/AzureSubnetSelectorService.java (hunk 2 above)
Comment (gergopapi2) [reply cloud-azure/src/main/java/com/sequenceiq/cloudbreak/cloud/azure/subnet/selector/AzureSubnetSelectorService.java]: I changed it, thanks.

File: environment/src/main/java/com/sequenceiq/environment/network/service/SubnetIdProvider.java (hunk 3 above)
Comment (gergopapi2) [reply environment/src/main/java/com/sequenceiq/environment/network/service/SubnetIdProvider.java]: Yes, thanks for pointing that out! I added it.

**Pull request merged**